### ISSUE TITLE
TypeError when calling psi.canonical_form()

### ISSUE THREAD

**Issue by rdilip** [tenpy/tenpy#98]
For a particular model I'm working on, I noticed that for certain parameters I get the following error when calling psi.canonical_form() after running DMRG for a few sweeps.

```
`<ipython-input-489-24ebdee29ec9> in <module>
----> 1 psi.canonical_form()

tenpy/networks/mps.py in canonical_form(self, renormalize)
   1863             return self.canonical_form_finite(renormalize)
   1864         else:
-> 1865             self.canonical_form_infinite(renormalize)
   1866 
   1867     def canonical_form_finite(self, renormalize=True, cutoff=0.):

tenpy/networks/mps.py in canonical_form_infinite(self, renormalize, tol_xi)
   2009             self.norm *= np.sqrt(norm)
   2010         # make Gr diagonal to Wr
-> 2011         Wr, Gl = self._canonical_form_correct_right(i1, Gr, return_Gl_guess=True)
   2012         # find dominant left eigenvector
   2013         norm, Gl = self._canonical_form_dominant_gram_matrix(i1, True, tol_xi, Gl)

tenpy/networks/mps.py in _canonical_form_correct_right(self, i1, Gr, return_Gl_guess, eps)
   2699         self.set_B(i1, npc.tensordot(Kr, self.get_B(i1), axes=['vR', 'vL']))
   2700         if return_Gl_guess:
-> 2701             guess = npc.tensordot(Kr.scale_axis(self.get_SL(i1)**2, 1), Kl, axes=['vR', 'vL'])
   2702             return W, guess.iset_leg_labels(['vR*', 'vR'])
   2703         return W
```

The model that reproduces this behavior is a bit involved to construct, so I haven't been able to construct a simple MWE that makes this problem apparent (nor have I found an easy way to reproduce this using any of the models predefined in tenpy). The issue looks like psi.get_SL() in this case returns an npc Array, rather than an np.ndarray, and there is no `__pow__` method implemented for the npc Array class. I also attached a pickled version of the mps where this issue appears.

The issue seems solvable without details of the model (though if this assumption is incorrect, please let me know and I can attach an example that defines the model introducing this bug).

[error_dump.pkl.tar.gz](https://github.com/tenpy/tenpy/files/4191467/error_dump.pkl.tar.gz)

**Comment by Lschoon** [comment]
Is this a recurrence of #89? If it is, could you post a snippet of code showing your runtime script?

It would also be helpful if you could list the parameters at which this error occurs, as well as a set of parameters that do not exhibit the error.

**Comment by rdilip** [comment]
I think it a similar issue, where psi._S has npc.Array entries instead of numpy arrays. However, I haven't been able to work out why it only arises for certain parameters. Sorry about the code dump, but as I haven't been able to get the error to show up on different models, I also defined the model i'm using in the below code. 

It's worth noting that the issue occurs for more points at a higher bond dimension. I've commented in a couple of points where this isn't an issue, and one where it is, at the bottom of the script.

```python
import numpy as np
from tenpy.models.model import CouplingMPOModel, NearestNeighborModel
from tenpy.tools.params import get_parameter
from tenpy.networks.site import BosonSite
from tenpy.models.lattice import Lattice
from tenpy.networks.mps import MPS
import numpy as np
import pickle
from tenpy.algorithms import dmrg

class MyModelChain(CouplingMPOModel):
    def __init__(self, model_params):
        L = model_params.get("L", 12)
        filling = model_params.get("filling", 0.25)
        bc_MPS = model_params.get("bc_MPS", "finite") 
        bc = "open"
        if bc_MPS == "infinite":
            bc = "periodic"
        n_max = model_params.get("n_max", 1)
        conserve = model_params.get("conserve", "N")
        site = BosonSite(n_max, conserve, filling)
        lattice = TwoSiteChain(site = site, L=L,
                                bc_MPS = bc_MPS,
                                bc = bc)
        self.lattice = lattice
        model_params["lattice"] = lattice
        CouplingMPOModel.__init__(self, model_params)

    def init_sites(self, model_params):
        n_max = get_parameter(model_params, 'n_max', 3, self.name)
        filling = get_parameter(model_params, 'filling', 0.25, self.name)
        conserve = get_parameter(model_params, 'conserve', 'N', self.name)
        print(self.name + ": set conserve to", conserve)
        site = BosonSite(Nmax=n_max, conserve=conserve, filling=filling)
        return site

    def init_terms(self, model_params):
        t = get_parameter(model_params, 't', 1., self.name, True)
        U = get_parameter(model_params, 'U', 0., self.name, True)
        V = get_parameter(model_params, 'V', 0., self.name, True)

        for u in range(len(self.lat.unit_cell)):
            if u % 2 == 0:
                self.add_onsite(2. * V, u, "N")
            if u % 2 == 1:
                self.add_onsite(2. * V, u, "NN")

        for u1, u2, dx in self.lat.nearest_neighbors:
            self.add_coupling(U, u1, 'N', u2, 'N', dx)
            self.add_coupling(-t, u1, 'Bd', u2, 'B', dx)
            self.add_coupling(-np.conj(t), u2, 'Bd', u1, 'B', -dx)  # h.c.
        for u1, u2, dx in self.lat.next_nearest_neighbors:
            if u1 % 2 == 0:
                self.add_coupling(U, u1, 'N', u2, 'N', dx)
            if u1 % 2 == 1:
                self.add_coupling(-2. * V, u1, "N", u2, "N", dx)
        for u1, u2, dx in self.lat.next_next_nearest_neighbors:
            self.add_coupling(-V, u1, "N", u2, "N", dx)

class TwoSiteChain(Lattice):
    """A chain of L sites. Unit cell is size 2.

    Parameters
    ----------
    L : int
        The lenght of the chain.
    site : :class:`~tenpy.networks.site.Site`
        The local lattice site. The `unit_cell` of the :class:`Lattice` is just ``[site]``.
    **kwargs :
        Additional keyword arguments given to the :class:`Lattice`.
        `[[next_]next_]nearest_neighbors` are set accordingly.
    """

    def __init__(self, L, site, **kwargs):
        kwargs.setdefault('nearest_neighbors',
                [(0, 1, np.array([0,])), (1, 0, np.array([1,]))])

        kwargs.setdefault('next_nearest_neighbors', 
                [(0, 0, np.array([1,])), (1, 1, np.array([1,]))])

        kwargs.setdefault('next_next_nearest_neighbors',
                [(0, 1, np.array([1,])), (1, 0, np.array([2,]))])
        # and otherwise default values.
        Lattice.__init__(self, [L//2], [site, site], **kwargs)

def run_simulation(**kwargs):
    data = {}
    v = kwargs.get("v", 0.0)
    chi_max = kwargs.get("chi_max", 500)
    U = kwargs.get("U", 10.0)
    t = kwargs.get("t", 1.0)
    L = 4

    model_params = dict(conserve="N",
                        n_max = 1,
                        t = t,
                        U = U,
                        V = v,
                        L = L,
                        bc_MPS = "infinite",
                        verbose = 1)

    dmrg_params = {"lanczos_params": {"N_min": 5, "N_max": 20, "reortho":True, "N_cache":22},
                    "mixer": True,
                    "chi_list": {0: 9, 10: 29, 20: 39, 30: 49, 40:59, 50:69, 60:73, 70: chi_max},
                    "trunc_params": {"svd_min": 1.e-10},
                    "mixer_params":
                        {"amplitude": 1.e-3, "decay":1.1, "disable_after":100},
                    "max_sweeps":200,
                    "verbose": 1,
                    "update_env": 0,
                    "max_E_err": 1.e-10,
                    "active_sites": 2
                    }
    
    M = MyModelChain(model_params)
    pstate = [1, 0, 0, 0]
    psi = MPS.from_product_state(M.lat.mps_sites(), pstate, M.lat.bc_MPS)

    try:
        info = dmrg.run(psi, M, dmrg_params)
    except:
        print("Hit an error")
        with open(f"error_dump_v_{round(v,3)}.pkl", "wb+") as f:
            pickle.dump(psi, f)
        sys.exit(0)

if __name__ == '__main__':
    U = 10.
    t = 1.0
    chi_max = 500
    v = 0.9175 # Throws error
    v = 0.9 # Does not throw error
    v = 1.0 # Does not throw error
    print(v)

    run_simulation(**dict(v=v, chi_max=chi_max, U=U, t=t))
```

**Comment by Lschoon** [comment]
The error is reproduced on my machine.

A workaround: if I tell the mixer to stop slightly earlier than in your original code (``"disable_after":90`` rather than ``100``), the error seems to disappear (please confirm). 

However, the problem should not occur in the first place, so I'm looking into why it happens now.

**Comment by Lschoon** [comment]
Alright, I think I have diagnosed the problem. If, coincidentally, the mixer is disabled in the same set of sweeps (*) as DMRG reaches convergence, no final sweeps are performed and the algorithm is terminated while ``psi._S`` are not yet in the correct form. Changing ``"disable_after"`` ,``"min_sweeps"`` or ``"N_sweeps_check"`` all result in the error disappearing.

Potential fixes for the underlying problem are:
- A reordering of the convergence criteria in `DMRGEngine.run()`` (potentially an extensive change)
- Removing the check for ``mixer is not None`` in ``Sweep.mixer_cleanup()``. Come to think of it, it looks like this check prevents ``mixer_cleanup()`` from ever acting, as by the time this method gets called, the mixer should already have been disabled. A quick test shows that removing this check removes the problem, so this has my preference.
@jhauschild, do you see any problems with removing this check that I missed?

(*) By 'set of sweeps', I mean the number of sweeps ``N_sweeps_check`` between checks for convergence

**Comment by rdilip** [comment]
This is probably redundant at this point, but I wanted to confirm that all the changes you mention (that lead to the convergence being out of sync with the mixer disabled) solve this problem.

**Comment by jhauschild** [comment]
I think completely removing the check is probably annoying for people who want to fine-tune the number of sweeps, e.g. if each of them is very expensive in terms of computation time.
The correct fix should be to replace the condition
`self.mixer is None` with what we actually want to check, i.e.,
`any([isinstance(self.psi.get_SL(i), npc.Array) for i in range(self.psi.L)])`
or `any([self.psi.get_SL(i).ndim > 1 for i in range(self.psi.L)])` [1].

Also, doing another sweep without optimization is just a simple hack/workaround,
what we actually should do is to perform SVDs `U, S1D, V = npc.svd(S)` and multiply the `U`, `V` into `psi` and the environments. But I guess that's another issue...

[1] This works since the push I've just made, adding `Array.ndim` as alias for `Array.rank` to simplify the comparison to numpy arrays.

**Comment by Lschoon** [comment]
Apologies for being silent on this for a while. `mixer_cleanup()` currently gets called only after all sweeps have taken place and should thus not affect a fine tuning of the number of sweeps. Do you expect this to change in the future?

**Comment by jhauschild** [comment]
Indeed, it get's called only after all the regular DMRG sweeps, and I don't expect that to change.

However, as of now, it simply does another 'environment_sweep', so it is a sweep itself. As we have seen in #95, this can be a bad thing one might want to avoid.
Also, it's completely unnecesary to do the extra environment sweep if everything is converged already and no mixer was used in the last sweep - which is what I expect in most situations.

All I'm saying is that we should keep the check around, and replace it with what we actually want to check - whether the singular values are a 1D numpy array or 2D npc Arrays.
